alf/algorithms/algorithm.py

```diff
@@ -309,6 +314,247 @@ def convert_train_state_to_predict_state(self, state):
                                        self._predict_state_spec)
         return state
 
+    @common.add_method(nn.Module)
+    def state_dict(self,
+                   destination=None,
+                   prefix='',
+                   keep_vars=False,
+                   visited=None):
+        """Get state dictionary recursively, including both model state
+        and optimizers' state (if any). It can handle a number of special cases:
+            1) graph with cycle: save all the states and avoid infinit loop
+            2) parameter sharing: save only one copy of the shared module/param
+            3) optimizers: save the optimizers for all the (sub-)algorithms
+
+        Args:
+            destination (OrderedDict): the destination for storing the state
+            prefix (str): a string to be added before the name of the items
+                (modules, params, algorithms etc) as the key used in the
+                state dictionary
+            keep_vars (bool): detach from params and buffers if False
+            visited (set): a set keeping track of the visited objects
```

Comment:
This `visited` is basically an internal argument that's never used by a normal user. A better way is to remove this argument and define another child function inside for the recursion.